### PR TITLE
Fix assignment validation in parser_append_binary_token

### LINKED ISSUE #3589: Memory Access Violation in ecma_compare_ecma_non_direct_strings
###### JerryScript revision
9b393ee2ea07ef4e7971e72637e46923dfc15e6f

###### Build platform
Ubuntu 18.04.4 LTS (Linux 4.15.0-76-generic x86_64)

###### Build steps
```
./tools/build.py --clean --compile-flag=-fsanitize=address \
--compile-flag=-m32 --compile-flag=-fno-omit-frame-pointer \
--compile-flag=-fno-common --compile-flag=-g \
--strip=off --system-allocator=on --logging=on \
--linker-flag=-fuse-ld=gold --error-messages=on --profile=es2015-subset
```

###### Output

```
AddressSanitizer:DEADLYSIGNAL
=================================================================
==16395==ERROR: AddressSanitizer: SEGV on unknown address 0x5f5e0ff4 (pc 0x0813fa83 bp 0xffc3b0c8 sp 0xffc3b0b0 T0)
==16395==The signal is caused by a READ memory access.

    #0 0x813fa82 in ecma_compare_ecma_non_direct_strings /root/work/libfuzzer/jerryscript/jerry-core/ecma/base/ecma-helpers-string.c:1819:20
    #1 0x81476d3 in ecma_find_named_property /root/work/libfuzzer/jerryscript/jerry-core/ecma/base/ecma-helpers.c:707:13
    #2 0x8181593 in ecma_op_object_get_own_property /root/work/libfuzzer/jerryscript/jerry-core/ecma/operations/ecma-objects.c:236:33
    #3 0x818209c in ecma_op_object_get_property /root/work/libfuzzer/jerryscript/jerry-core/ecma/operations/ecma-objects.c:354:32
    #4 0x818209c in ecma_op_object_has_property /root/work/libfuzzer/jerryscript/jerry-core/ecma/operations/ecma-objects.c:409:30
    #5 0x8235efa in ecma_op_put_value_lex_env_base /root/work/libfuzzer/jerryscript/jerry-core/ecma/operations/ecma-get-put-value.c:275:13
    #6 0x81e7e8d in vm_loop /root/work/libfuzzer/jerryscript/jerry-core/vm/vm.c:3874:43
    #7 0x81de804 in vm_execute /root/work/libfuzzer/jerryscript/jerry-core/vm/vm.c:4194:37
    #8 0x81de12f in vm_run /root/work/libfuzzer/jerryscript/jerry-core/vm/vm.c:4302:10
    #9 0x81de3b7 in vm_run_global /root/work/libfuzzer/jerryscript/jerry-core/vm/vm.c:320:10
    #10 0x81235c7 in jerry_run /root/work/libfuzzer/jerryscript/jerry-core/api/jerry.c:593:24
    #11 0x811b483 in main /root/work/libfuzzer/jerryscript/jerry-main/main-unix.c:748:23
    #12 0xf7cd4e80 in __libc_start_main (/lib32/libc.so.6+0x18e80)
    #13 0x8071eb1 in _start (/root/work/libfuzzer/jerryscript/build/bin/jerry+0x8071eb1)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV /root/work/libfuzzer/jerryscript/jerry-core/ecma/base/ecma-helpers-string.c:1819:20 in ecma_compare_ecma_non_direct_strings
==16395==ABORTING
```

### PR DESCRIPTION
**Pull request by szilagyiadam**
Fixes #3589

JerryScript-DCO-1.0-Signed-off-by: Adam Szilagyi aszilagy@inf.u-szeged.hu

### CODE REVIEW

File: tests/jerry/es2015/regression-test-issue-3589.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+try {
+  [this,000000000,this,99999999=9999999];
+  assert(false);
+} catch (e) {
+  assert(e instanceof ReferenceError);
```
Comment (galpeter) [inline tests/jerry/es2015/regression-test-issue-3589.js]: Is it ok to expect a ReferenceError? It should be a SyntaxError (as assigning a value to a constant is not valid)

Comment (rerobika) [reply tests/jerry/es2015/regression-test-issue-3589.js]: That's also an es5.1 bug, so the tc should be moved as well. However ES6 static semantic is not implemented  yet, therefore I'd suggest to keep the ES5.1 solution for now.

File: jerry-core/parser/js/js-parser-expr.c [hunk 2]
```diff
@@ -2180,10 +2179,19 @@ parser_append_binary_single_assignment_token (parser_context_t *context_p, /**<
   }
   else if (context_p->last_cbc_opcode == CBC_PUSH_THIS_LITERAL)
   {
-    context_p->last_cbc_opcode = CBC_PUSH_THIS;
-    parser_flush_cbc (context_p);
-    parser_stack_push_uint16 (context_p, context_p->last_cbc.literal_index);
-    parser_stack_push_uint8 (context_p, assign_ident_opcode);
+    if (context_p->last_cbc.literal_type != LEXER_IDENT_LITERAL)
```
Comment (rerobika) [inline jerry-core/parser/js/js-parser-expr.c]: I'd move this check into `parser_check_invalid_assign`.

**Comment by rerobika** [comment]
Please rebase the PR to the latest master.

**Pull request merged**